design/010-pushsecret-metadata.md

```diff
@@ -0,0 +1,144 @@
+```yaml
+---
+title: PushSecret metadata
+version: v1alpha1
+authors: Moritz Johner
+creation-date: 2023-08-25
+status: draft
+---
+```
+
+# PushSecret Metadata
+
+[#2600](https://github.com/external-secrets/external-secrets/pull/2600) introduced a new feature that allows users to pass arbitrary `metadata` to the provider.
+
+The data is arbitrary json/yaml and can be anything.
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: PushSecret
+metadata:
+  name: pushsecret-example
+spec:
+  # ...
+  data:
+    - match:
+        secretKey: key1
+        remoteRef:
+          remoteKey: test1
+      metadata:
+        annotations:
+          key1: value1
+        labels:
+          key1: value1
+
+```
+
+Here is a overview of current implementations of PushSecret metadata:
+
+```yaml
+# AWS Parameter Store
+# more to come in https://github.com/external-secrets/external-secrets/pull/3581
+parameterStoreType: "..."
+parameterStoreKeyID: "..."
+```
+
+```yaml
+# GCP Secrets Manager
+labels: {}
+annotations: {}
+```
+
+```yaml
+# AWS Secrets Manager
+secretPushFormat: "..."
+```
+
+## Problem Description
+
+We will never be able to make disruptive changes, we can only append to the existing structure.
+
+**Why is that a problem?**
+
+It limits our ability to fix mistakes that have been merged and released. Having an `apiVersion` field would allow us decode the metadata differently and apply the appropriate logic in a code branch. 
+
+This would simplify fixing simple mis-nomers or doing large-scale refactorings in the future. 
+
+ESO is a community based project and relies on contributions from different backgrounds and experience levels. As a result, the approach and perspective to a solution highly depends
+on the contributor and the reviewer. We will eventually have to align the structure or naming of metadata across providers once we see patterns emerge.
+
+## Proposed Solution
+
+I would propose to wrap the unstructured metadata in a Kubernetes *alike* resource containing an `apiVersion`, `kind` and `spec`. 
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: PushSecret
+metadata:
+  name: pushsecret-example
+spec:
+  # ...
+  data:
+    - match:
+        secretKey: key1
+        remoteRef:
+          remoteKey: test1
+      metadata:
+        apiVersion: kubernetes.external-secrets.io/v1alpha1
+        kind: PushSecretMetadata
+        spec:
+          sourceMergePolicy: Merge
+          targetMergePolicy: Merge
+          labels:
+            color: red
+          annotations:
+            yes: please
+```
+
+**PROS**
+- familiar structure for Kubernetes users, other projectes use that pattern already
```

Comment:
In general I prefer this, but I wonder how we would get early feedback about it before committing, that would be great :thinking: 

I guess versioning it (v1alpha1) as proposed already fives us that liberty to wait and check and get feedback. I just don't the "silence is golden" in this context as we have in kunernetes and other communities. Wish we had a better way to check community sentiment that is less passive and timely

Reply:
Also, this would somewhat break existing usages of the metadata but hopefully, not in an aggressive way. For now, I think it's mostly AWS and Secret format that are actively using this field, right?

Reply:
I am in favor of this proposal. It might add further clarity to include an example PushSecretMetadata for one of the non-kubernetes providers, say AWS Secrets Manager.

Reply:
> Wish we had a better way to check community sentiment that is less passive and timely

We'll get feedback (if any) once it is implemented, released and documented. We need the metadata, that is a fact. How we want to structure it is up to us, no users will ever approach us and say "we need versioning for metadata" :D

Reply:
Sounds good to me.

Reply:
Thanks much for the additional PushSecretMetadata examples.